UCBChangesToMorphic.js

```diff
@@ -0,0 +1,306 @@
+        while (!trg.droppedSVG) {
+            trg = trg.parent;
+        }
+        pic.onload = () => {
+            trg.droppedSVG(pic, aFile.name);
+            bulkDrop();
+        };
+        frd = new FileReader();
+        frd.onloadend = (e) => pic.src = e.target.result;
+        frd.readAsDataURL(aFile);
+    }
+
+    function readImage(aFile) {
+        var pic = new Image(),
+            frd = new FileReader(),
+            trg = target;
+        while (!trg.droppedImage) {
+            trg = trg.parent;
+        }
+        
+        /* UCB Script Pic Edit:
+            *  - Added new process for accepting images dropped on script window
+            *  - Images are searched for XML metadata, and if present, users are able to:
+            *      1. Process the script as an XML file and load script
+            *      2. Process the script as an image and load it into the costumes
+            *  - Base code credit given do Snap-Forum user Dardoro
+            * 
+            * 2/26 Changes:
+            *  - Attempting to break up custom block definitions and scripts into 2 drops
+            *  - Added asyncronous call to ide.droppedText() for both drops
+            */
+        function arr2Str(arr) { 
+            return arr.reduce((acc, b) => acc + String.fromCharCode(b), "");
+        };
+        
+        let ide = world.children[0],
+            url = event.dataTransfer?.getData( "text/uri-list"),
+            file = event.dataTransfer?.files?.[0],
+            menu = new MenuMorph(this, "Process Dropped Object");
+
+        pic.onload = () => {
+            canvas = newCanvas(new Point(pic.width, pic.height), true);
+            canvas.getContext('2d').drawImage(pic, 0, 0);
+
+            (async () => {
+                if (!file && url) {
+                    file = await fetch("https://api.allorigins.win/raw?url="+url)
+                };
+                let buff = new Uint8Array(await file?.arrayBuffer()),
+                    strBuff = arr2Str(buff),
+                    blocks;
+                
+                strBuff.includes("Snap\tBlocks\tEmbedded") 
+                    ? blocks = decodeURIComponent(escape((strBuff)?.split("Snap\tBlocks\tEmbedded")[1]))
+                    : blocks = null;
+
+                if (blocks) { 
+                    menu.addItem(
+                        "Interpret script pic as blocks.", 
+                        () => {
+                            if (blocks.includes("Custom\tBlock\tDefinitions")) {
+                                let customsSplit = blocks.split("Custom\tBlock\tDefinitions");
+                                await ide.droppedText(customsSplit[0], file.name, ""); // drop definitions on first
+                                await ide.droppedText(customsSplit[1], file.name, ""); // drop scripts on afterwards
+                            } else {
+                                ide.droppedText(blocks, file.name, "");
+                            } 
```

Comment:
To be able to properly decode the XML inside the images, I tried two things. At first, I thought that simply concatenating the custom block definitions with other scripts would properly bring everything back. However, there were messages about missing block definitions. 

To try to get around that, I separated the definitions and scripts by a delimiter. The thought process behind this was that I'd be able to drop the custom block definitions onto the `ide` then drop the scripts afterwards so there would be no errors. However, I found that only the second drop is successful and the other is disregarded. Even after adding `async/await` calls to the function calls, only the second call succeeded.

Any thoughts on what to do?

Reply:
For example, the following would be embedded in the PNG below:
`<blocks app="Snap! 7dev, https://snap.berkeley.edu" version="2"><block-definition s="bar" type="command" category="other"><header></header><code></code><translations></translations><inputs></inputs></block-definition><block-definition s="foo" type="command" category="other"><header></header><code></code><translations></translations><inputs></inputs></block-definition></blocks>Custom	Block	Definitions<script><block s="setHeading"><l>90</l></block><custom-block s="bar"></custom-block><block s="changeXPosition"><l>10</l></block><custom-block s="foo"></custom-block></script>`

I left the definitions empty for simplicity, but the same applies with blocks within them.

![setHeading (19)](https://user-images.githubusercontent.com/72178487/155875045-37ac9b13-f3a3-47f5-b2d1-398bdcd0d821.png)

Reply:
I've got mediocre success with this direct synchronous import
```
ide.rawOpenBlocksString( customsSplit[0], file.name, true); // import custom blocks
ide.openScriptString( customsSplit[1], file.name, false); // drop scripts on afterwards

```
Custom blocks are imported OK,  but `"<script>"` import seems to be flawed, in this version, regardless of the source - XML or PNG.

Reply:
Thank you for the functions! I just added them in a new commit. I don't have an idea of how to get around the `Undefined` blocks, but @jmoenig @brianharvey @cycomachead if you have any ideas they'd be greatly appreciated :)

Reply:
Makeshift workaround, till import of the script will be fixed in Snap! 
```
delete ide.serializer.scene.stage;//workaround, force loading script in "ide.stage" context
ide.openScriptString(customsSplit[1], file.name, false);
```